API.md

```diff
@@ -0,0 +1,319 @@
+
+An **OpenSCD menu plugin** is an OpenSCD plugin with an additional `run()` instance method which is called when the plugin's entry is clicked in OpenSCD's menu. It is continuously rendered into the app and is expected to normally not display any content of its own. It is meant for one-shot editing tasks or tasks that always run in the background (like validation).
+
+An **OpenSCD editor plugin** is a modal OpenSCD plugin that is only rendered as long as the user has its tab selected in OpenSCD's tab bar. It is meant for rendering the main part of OpenSCD's user interface.
+
+The **OpenSCD core API** is:
+- the way in which OpenSCD core communicates relevant data to the plugins and
+- the way in which plugins communicate user intent to OpenSCD core.
+- the way in which OpenSCD sets CSS fonts and colors for plugins
+
+## Communicating data to plugins
+
+OpenSCD core communicates the data necessary for editing SCL documents by setting the following [properties](https://developer.mozilla.org/en-US/docs/Glossary/Property/JavaScript) on the plugin's [DOM Element](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement):
+
+
+```typescript=
+export default class Plugin extends LitElement {
+  docs: Record<string, XMLDocument> = {};
+  doc?: XMLDocument;
+  docName?: string;
+  editable: string[] = [
+    'cid',
+    'icd',
+    'iid',
+    'scd',
+    'sed',
+    'ssd',
+  ];
+  history: LogEntry[];
+  editCount: number = 0;
+  plugins: { menu: Plugin[], editor: Plugin[] }[];
+  locale: string = 'en';
+}
+
+/** Helper types exported by OpenSCD core **/
+
+type LogEntry = {
+  redo: Edit; // `Edit` defined below
+  undo: Edit;
+  title?: string
+}
+
+type Plugin = {
+  name: string;
+  translations?: Record<string, string>;
+  src: string;
+  icon: string;
+  requireDoc?: boolean; // disable plugin if no doc is opened
+}
+```
+
+### `docs`
+
+`docs` is set to an object mapping `string` keys (document names) to `XMLDocument` values.
+
+### `docName`
+The name of the `XMLDocument` currently being edited.
+
+### `doc`
+The `XMLDocument` currently being edited.
+
+### `editable`
+Filename extensions of user-editable documents.
+
+### `history`
+History of edits done to `doc`.
+
+### `editCount`
+
+Index of the current log entry in the `history`. Is incremented with every new edit and updated on undo/redo. 
+
+### `plugins`
+
+Arrays of `Plugin` objects describing the currently loaded `menu` and `editor` plugins, respectively.
+
+### `locale`
+
+Selected language (IETF language tag).
+
+## Communicating user intent to OpenSCD core
+
+Plugins communicate user intent to OpenSCD core by dispatching the following [custom events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent):
+
+### `EditEventV2`
+
+The **edit event** allows a plugin to describe the changes it wants to make to the current `doc`.
```

Comment:
Last comment I promise! 

Is an edit event sometimes unsuccessful? Should our API allow a positive assertion of success/failure, like an exit code? Or is this unnecessary? (similar questions around our library development in scl-lib have been discussed recently and I would be interested in thoughts on this).

Reply:
Sadly, [dispatching an event only returns whether the event was cancelled, according to the Web API](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent#return_value). So the only feedback a plugin gets for dispatching an event is a change (or no change) to the edit `history`, which as of this API draft would be fully exposed to plugins through a property.

I'm open to other suggestions for convenient feedback mechanisms. Simply adding an optional callback to the event which will be called by core upon completion of the edit (successful or otherwise) along with some status code seems a bit ugly and cumbersome, but would be a last resort stop-gap measure for use cases for which the above updating of the `history` property would not suffice.

What are the use cases that you see for this feedback mechanism, @danyill and @JakobVogelsang ?

Reply:
For me it is again mainly when developing plugins. You simply do not know if the code is working correct, and then feedback mechanisms are important. A good indicator could be feedback on the edits send. I am not sure though if it is needed in production. Also I think this is something a plugin author could take care as well. When edits are sent it is quite predictable what the core component will do.

Reply:
I tend to agree, if a plugin sends through sane edits then core can generally be relied on. 

Then it becomes the task of users of core to send through correct edits in the first place. So perhaps an "exit code" is more a requirement on a downstream library or the responsibility of an individual plugin.

For the DOM API I note that `insertBefore` returns the inserted node whereas `setAttribute` doesn't return anything but can have exceptions which can be managed.

Should plugins be able to listen to an event to indicate that something has gone wrong during the edit process? (either the node not being inserted or an exception occurring)? I'm thinking something to do with the catch block: 

https://github.com/OpenEnergyTools/open-scd-core/blob/3f9aacb94eec216d6a27b9407fa1dc06714950ea/foundation/edit-event.ts#L73-L92

I am not knowledgeable about API design and error handling so I'll be happy for this comment to be passed over 😉

Reply:
Hope this is good enough for our use cases. If not, we can open a further issue to extend `Edit`s with return-code callbacks or the like at a later point.